core/src/test/java/org/apache/iceberg/TestMetrics.java

```diff
@@ -352,14 +327,34 @@ public void testMetricsForNullColumns() throws IOException {
     Record secondRecord = GenericRecord.create(schema);
     secondRecord.setField("intCol", null);
 
-    InputFile recordsFile = writeRecords(schema, firstRecord, secondRecord);
-
-    Metrics metrics = getMetrics(recordsFile);
+    Metrics metrics = getMetrics(schema, firstRecord, secondRecord);
     Assert.assertEquals(2L, (long) metrics.recordCount());
-    assertCounts(1, 2L, 2L, metrics);
+    assertCounts(1, 2L, 2L, 0L, metrics);
     assertBounds(1, IntegerType.get(), null, null, metrics);
   }
 
+  @Test
+  public void testMetricsForNaNColumns() throws IOException {
```

Comment:
There are a few cases to consider with `NaN` values because comparison with `NaN` is always false.

Here are a couple of implementations that have issues because they use comparison without checking for NaN:

```java
// max is NaN for values [NaN, 1.0, 1.1]
float max = null;
for (value : values) {
  if (max == null || max < value) {
    max = value;
  }
}

// max is NaN if any value is NaN
float max = null;
for (value : values) {
  max = (max != null && max >= value) ? max : value;
}
```

Because the failure cases are different, I think we should test a few different cases:
* A column starts with NaN
* A column contains NaN in the middle
* A column ends with NaN

Reply:
Thanks for pointing this out! I guess this tests more about boundary than of NaN count, but it would be very helpful when we start to exclude NaN from upper/lower bounds. Will update.